File: day2/main.py
def parse(l):
    name, content = l.split(":")
    i = int(name.split(" ")[1])
    g = []    
    for s in content.split(";"):
        v = {"red": 0, "green": 0, "blue": 0}
        for item in s.split(","):
            item = item[1:]
            n, col = item.split(" ")
            col = col.rstrip()
            assert v[col] == 0
            v[col] = int(n)
        g.append(v)
    return i, g

def possible(g):
    return all(v["red"]<=12 and v["green"]<=13 and v["blue"]<=14 for v in g)

def part1(gs):
    total = 0
    for i,g in gs:
        if possible(g):
            total += i
    return total

def part2(gs):
    total = 0
    for _, g in gs:
        p = max(v["red"] for v in g) * max(v["blue"] for v in g) * max(v["green"] for v in g)
        total += p
    return total

if __name__ == "__main__":
    with open("./day2/input.txt") as f:
        ls = f.readlines()
    gs = []
    for l in ls:
        i, g = parse(l)
        gs.append((i,g))
    print(part1(gs))
    print(part2(gs))

File: day2/main.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <list>

using namespace std;

struct Draw { int r; int g; int b; };

bool possible(list<Draw> &game) {
    for (const Draw& v : game)
        if (v.r > 12 || v.g > 13 || v.b > 14) return false;
    return true;
}

int power(list<Draw> &game) {
    Draw vmax{0, 0, 0};
    for (const Draw& v : game) {
        if (v.r > vmax.r) vmax.r = v.r;
        if (v.b > vmax.b) vmax.b = v.b;
        if (v.g > vmax.g) vmax.g = v.g;
    }
    return vmax.r * vmax.g * vmax.b;
}

void trim(string& s) {
    while (!s.empty() && s[0] == ' ') s.erase(0,1);
}

void parseline(const string& line, list<Draw>& game, int& game_id) {
    string name = line.substr(0, line.find(':'));
    string content = line.substr(line.find(':')+2);
    game_id = stoi(name.substr(name.find(' ')+1));

    stringstream ss(content);
    string s;
    while (getline(ss, s, ';')) {
        Draw v{0, 0, 0};
        stringstream is(s);
        string c;
        while (getline(is, c, ',')) {
            trim(c);
            int n = stoi(c.substr(0,c.find(' ')));
            string col = c.substr(c.find(' ')+1);
            if (!col.empty() && col.back() == '\n') col.pop_back();
            if (col == "red") v.r = n;
            if (col == "green") v.g = n;
            if (col == "blue") v.b = n;
        }
        game.emplace_back(v); 
    }
}

int main() {
    ifstream infile("input.txt");
    string line;
    int total1 = 0;
    int total2 = 0;
    while (getline(infile,line)) {
        list<Draw> game;
        int game_id;
        parseline(line, game, game_id);
        if (possible(game)) {
            total1 += game_id;
        }
        total2 += power(game);
    }
    cout << total1 << endl;
    cout << total2 << endl;
}